File: dart/lib/card.dart
// ignore_for_file: constant_identifier_names

/// Enum for card suits and their most basic methods.
/// The actual numbers are important here: change them and you will break
/// some tests and compatibility with the Rust code.
enum Suit {
  None, // 0
  Club, // 1
  Diamond, // 2
  Heart, // 3
  Spade; // 4

  /// Create one from text character or unicode symbol.
  static Suit fromChar(String c) {
    switch (c) {
      case 'c':
      case '♣':
        return Club;
      case 'd':
      case '♦':
        return Diamond;
      case 'h':
      case '♥':
        return Heart;
      case 's':
      case '♠':
        return Spade;
      default:
        return None;
    }
  }

  /// Render as text
  String toChar() {
    if (index < 0 || index > 4) return '?';
    return ['?', 'c', 'd', 'h', 's'][index];
  }

  /// Render as unicode symbol
  String toUnicode() {
    if (index < 0 || index > 4) return '?';
    return ['?', '♣', '♦', '♥', '♠'][index];
  }

  /// Full English name
  String get name {
    if (index < 0 || index > 4) return "?";
    return ["?", "club", "diamond", "heart", "spade"][index];
  }

  /// Full English plural name, not really needed but for consistency
  String get plural {
    if (index < 0 || index > 4) return "?";
    return ["?", "clubs", "diamonds", "hearts", "spades"][index];
  }

  /// For consistency with rank, not really needed
  String get article {
    return "a";
  }
}

/// Enum for card ranks and their basic methods. As with suits, do not
/// change the numbers.
enum Rank {
  None, // 0
  LowAce, // 1
  Deuce, // 2
  Trey, // 3
  Four, // 4
  Five, // 5
  Six, // 6
  Seven, // 7
  Eight, // 8
  Nine, // 9
  Ten, // 10
  Jack, // 11
  Queen, // 12
  King, // 13
  Ace, // 14
  Knight; // 15

  /// Create one from text character.
  static Rank fromChar(String c) {
    switch (c) {
      case '1': // Extension for testing never really used
        return LowAce;
      case '2':
        return Deuce;
      case '3':
        return Trey;
      case '4':
        return Four;
      case '5':
        return Five;
      case '6':
        return Six;
      case '7':
        return Seven;
      case '8':
        return Eight;
      case '9':
        return Nine;
      case 'T':
        return Ten;
      case 'J':
        return Jack;
      case 'Q':
        return Queen;
      case 'K':
        return King;
      case 'A':
        return Ace;
      case 'C':
        return Knight;
      default:
        return None;
    }
  }

  /// Render as text
  String toChar() {
    if (index < 0 || index > 15) return '?';
    return [
      '?',
      'A',
      '2',
      '3',
      '4',
      '5',
      '6',
      '7',
      '8',
      '9',
      'T',
      'J',
      'Q',
      'K',
      'A',
      'C'
    ][index];
  }

  /// Full English name
  String get name {
    if (index < 0 || index > 15) return "?";
    return [
      "?",
      "ace",
      "deuce",
      "trey",
      "four",
      "five",
      "six",
      "seven",
      "eight",
      "nine",
      "ten",
      "jack",
      "queen",
      "king",
      "ace",
      "knight"
    ][index];
  }

  /// Full English plural name (to deal with "sixes")
  String get plural {
    if (index < 0 || index > 15) return "?";
    return [
      "?",
      "aces",
      "deuces",
      "treys",
      "fours",
      "fives",
      "sixes",
      "sevens",
      "eights",
      "nines",
      "tens",
      "jacks",
      "queens",
      "kings",
      "aces",
      "knights"
    ][index];
  }

  /// Indefinite article (to deal with "an eight, an ace")
  String get article {
    if (index == 1 || index == 8 || index == 14) return "an";
    return "a";
  }
}

/// Enum for cards and their basic methods. Numbers *very* important here,
/// not just for compatibility but specific algorithms will fail if changed.
enum Card {
  None, // 0
  WhiteJoker, // 1
  BlackJoker, // 2
  Joker, // 3
  LowAceOfClubs, // 4
  LowAceOfDiamonds, // 5
  LowAceOfHearts, // 6
  LowAceOfSpades, // 7
  DeuceOfClubs, // 8
  DeuceOfDiamonds, // 9
  DeuceOfHearts, // 10
  DeuceOfSpades, // 11
  TreyOfClubs, // 12
  TreyOfDiamonds, // 13
  TreyOfHearts, // 14
  TreyOfSpades, // 15
  FourOfClubs, // 16
  FourOfDiamonds, // 17
  FourOfHearts, // 18
  FourOfSpades, // 19
  FiveOfClubs, // 20
  FiveOfDiamonds, // 21
  FiveOfHearts, // 22
  FiveOfSpades, // 23
  SixOfClubs, // 24
  SixOfDiamonds, // 25
  SixOfHearts, // 26
  SixOfSpades, // 27
  SevenOfClubs, // 28
  SevenOfDiamonds, // 29
  SevenOfHearts, // 30
  SevenOfSpades, // 31
  EightOfClubs, // 32
  EightOfDiamonds, // 33
  EightOfHearts, // 34
  EightOfSpades, // 35
  NineOfClubs, // 36
  NineOfDiamonds, // 37
  NineOfHearts, // 38
  NineOfSpades, // 39
  TenOfClubs, // 40
  TenOfDiamonds, // 41
  TenOfHearts, // 42
  TenOfSpades, // 43
  JackOfClubs, // 44
  JackOfDiamonds, // 45
  JackOfHearts, // 46
  JackOfSpades, // 47
  QueenOfClubs, // 48
  QueenOfDiamonds, // 49
  QueenOfHearts, // 50
  QueenOfSpades, // 51
  KingOfClubs, // 52
  KingOfDiamonds, // 53
  KingOfHearts, // 54
  KingOfSpades, // 55
  AceOfClubs, // 56
  AceOfDiamonds, // 57
  AceOfHearts, // 58
  AceOfSpades, // 59
  KnightOfClubs, // 60
  KnightOfDiamonds, // 61
  KnightOfHearts, // 62
  KnightOfSpades; // 63

  /// Create one from integer value
  static Card fromInt(int v) {
    if (v < 1 || v > 63) return None;
    return Card.values[v];
  }

  /// Create one from rank and suit
  static Card fromRankSuit(Rank r, Suit s) {
    if (r == Rank.None || s == Suit.None) return None;
    return Card.values[(r.index << 2) | (s.index - 1)];
  }

  /// What's the rank of this card? Note that we use a shift on the number
  /// value; that's why the numbers are important. Note that jokers do not
  /// have rank or suit.
  Rank get rank {
    if (index < 4 || index > 63) return Rank.None;
    return Rank.values[index >> 2];
  }

  /// What's the rank of this card, with low aces made high?
  Rank get highRank {
    if (index < 4 || index > 63) return Rank.None;
    if (index < 8) return Rank.Ace;
    return Rank.values[index >> 2];
  }

  /// What's the suit of this card? Jokers have no suit.
  Suit get suit {
    if (index < 4 || index > 63) return Suit.None;
    return Suit.values[1 + (index & 3)];
  }

  /// Is this an actual card (and not None or an invalid number)?
  bool get isCard {
    return index > 0 && index < 64;
  }

  /// Is this card an ace?
  bool get isAce {
    return (index >= 4 && index <= 7) || (index >= 56 && index <= 59);
  }

  /// Is this card a joker?
  bool get isJoker {
    return (index >= 1 && index <= 3);
  }

  /// Is this card red? Note that while jokers have no suit, there is a
  /// black one and a red one.
  bool get isRed {
    if (3 == index) return true;
    if (index < 4 || index > 63) return false;
    return (index & 3) == 1 || (index & 3) == 2;
  }

  /// Is this card black? Note that while jokers have no suit, there is a
  /// black one and a red one.
  bool get isBlack {
    if (2 == index) return true;
    if (index < 4 || index > 63) return false;
    return (index & 3) == 0 || (index & 3) == 3;
  }

  /// Render to two-character text in standard format
  String toText() {
    if (!isCard) return "??";
    if (1 == index) return "Jw";
    if (2 == index) return "Jb";
    if (3 == index) return "Jk";
    return "${rank.toChar()}${suit.toChar()}";
  }

  /// Render to two-character text with unicode suit symbols
  String toUnicode() {
    if (!isCard) return "??";
    if (1 == index) return "Jw";
    if (2 == index) return "Jb";
    if (3 == index) return "Jk";
    return "${rank.toChar()}${suit.toUnicode()}";
  }

  /// Full English name
  String get fullName {
    if (!isCard) return "unknown";
    if (1 == index) return "white joker";
    if (2 == index) return "black joker";
    if (3 == index) return "joker";
    return "${rank.name} of ${suit.plural}";
  }

  /// Make high aces low, leave other cards alone
  static Card lowAceFix(Card c) {
    return Card
        .values[(c.index >= 56 && c.index <= 59) ? c.index - 52 : c.index];
  }

  /// Make low aces high, leave other cards alone
  static Card highAceFix(Card c) {
    return Card.values[(c.index >= 4 && c.index <= 7) ? c.index + 52 : c.index];
  }

  @override
  String toString() {
    return toText();
  }
}

/// Parse a string of card text into a sequence of cards. Whitespace is
/// ignored between cards, but is not allowed between rank and suit.
/// Cards may be enclosed in square brackets.
Iterable<Card> cardsFromText(String text) sync* {
  int state = 0;
  bool bracketAllowed = true;
  Suit s = Suit.None;
  Rank r = Rank.None;

  for (String c in text.runes.map((r) => String.fromCharCode(r))) {
    switch (state) {
      case 0: // skip whitespace or one [, look for rank
        if ("[" == c) {
          if (!bracketAllowed) {
            return;
          }
          bracketAllowed = false;
          break;
        }
        if (" " == c) {
          break;
        }
        bracketAllowed = false;

        if ("J" == c) {
          state = 1;
          break;
        }
        r = Rank.fromChar(c);
        if (Rank.None == r) {
          return;
        }
        state = 2;
        break;
      case 1: // got J, look for joker color or jack suit
        if ("k" == c || "r" == c) {
          yield Card.Joker;
          state = 0;
          break;
        }
        if ("b" == c) {
          yield Card.BlackJoker;
          state = 0;
          break;
        }
        if ("w" == c) {
          yield Card.WhiteJoker;
          state = 0;
          break;
        }
        r = Rank.Jack;
        s = Suit.fromChar(c);
        if (Suit.None == s) {
          return;
        }
        yield Card.fromRankSuit(r, s);
        state = 0;
        break;
      case 2: // got rank, look for suit
        s = Suit.fromChar(c);
        if (Suit.None == s) {
          return;
        }
        assert(Rank.None != r);
        yield Card.fromRankSuit(r, s);
        state = 0;
        break;
      default:
        assert(false);
    }
  }
}

File: deno/ts/card.ts
/**
 * @file cards.ts
 */

import * as Suit from "./suit.ts";
import * as Rank from "./rank.ts";

export type T = number;

export const None = 0;
export const WhiteJoker = 1;
export const RedJoker = 2;
export const Joker = 3;
export const LowAceOfClubs = 4;
export const LowAceOfDiamonds = 5;
export const LowAceOfHearts = 6;
export const LowAceOfSpades = 7;
export const DeuceOfClubs = 8;
export const DeuceOfDiamonds = 9;
export const DeuceOfHearts = 10;
export const DeuceOfSpades = 11;
export const TreyOfClubs = 12;
export const TreyOfDiamonds = 13;
export const TreyOfHearts = 14;
export const TreyOfSpades = 15;
export const FourOfClubs = 16;
export const FourOfDiamonds = 17;
export const FourOfHearts = 18;
export const FourOfSpades = 19;
export const FiveOfClubs = 20;
export const FiveOfDiamonds = 21;
export const FiveOfHearts = 22;
export const FiveOfSpades = 23;
export const SixOfClubs = 24;
export const SixOfDiamonds = 25;
export const SixOfHearts = 26;
export const SixOfSpades = 27;
export const SevenOfClubs = 28;
export const SevenOfDiamonds = 29;
export const SevenOfHearts = 30;
export const SevenOfSpades = 31;
export const EightOfClubs = 32;
export const EightOfDiamonds = 33;
export const EightOfHearts = 34;
export const EightOfSpades = 35;
export const NineOfClubs = 36;
export const NineOfDiamonds = 37;
export const NineOfHearts = 38;
export const NineOfSpades = 39;
export const TenOfClubs = 40;
export const TenOfDiamonds = 41;
export const TenOfHearts = 42;
export const TenOfSpades = 43;
export const JackOfClubs = 44;
export const JackOfDiamonds = 45;
export const JackOfHearts = 46;
export const JackOfSpades = 47;
export const KnightOfClubs = 48;
export const KnightOfDiamonds = 49;
export const KnightOfHearts = 50;
export const KnightOfSpades = 51;
export const QueenOfClubs = 52;
export const QueenOfDiamonds = 53;
export const QueenOfHearts = 54;
export const QueenOfSpades = 55;
export const KingOfClubs = 56;
export const KingOfDiamonds = 57;
export const KingOfHearts = 58;
export const KingOfSpades = 59;
export const AceOfClubs = 60;
export const AceOfDiamonds = 61;
export const AceOfHearts = 62;
export const AceOfSpades = 63;

export function suitOf(n: T): Suit.T {
    if (n < LowAceOfClubs || n > KnightOfSpades) return Suit.None;
    return (n & 3) + 1 as Suit.T;
}

export function rankOf(n: T): Rank.T {
    if (n < LowAceOfClubs || n > KnightOfSpades) return Rank.None;
    return (n >> 2) as Rank.T;
}

// Always return high aces. Good for indexing external tables, etc.
export function highRankOf(n: T): Rank.T {
    if (n < LowAceOfClubs || n > KnightOfSpades) return Rank.None;
    if (n < DeuceOfClubs) return Rank.Ace;
    return (n >> 2) as Rank.T;
}

export function fromRankSuit(r: Rank.T, s: Suit.T): T {
    if (r < 1 || r > 15) return None;
    if (s < 1 || s > 4) return None;
    return ((r << 2) | (s - 1)) as T;
}

/**
 * These are used internally for ensuring that the right values get into
 * hands from the chosen deck. You should need them as a library user.
 */
export function lowAceFix(n: T): T {
    return ((n >= 56 && n <= 59) ? n - 52 : n) as T;
}

export function highAceFix(n: T): T {
    return ((n >= 4 && n <= 7) ? n + 52 : n) as T;
}

/**
 * Various card classifiers.
 */
export function isCard(n: T): boolean {
    return n >= LowAceOfClubs && n <= AceOfSpades;
}

export function isAce(n: T): boolean {
    return (n >= LowAceOfClubs && n <= LowAceOfSpades) ||
        (n >= AceOfClubs && n <= AceOfSpades);
}

export function isJoker(n: T): boolean {
    return n >= WhiteJoker && n <= Joker;
}

export function isRed(n: T): boolean {
    if (n === RedJoker) return true;
    const s = suitOf(n);
    return Suit.Diamond === s || Suit.Heart === s;
}

export function isBlack(n: T): boolean {
    if (n === Joker) return true;
    const s = suitOf(n);
    return Suit.Club === s || Suit.Spade === s;
}

/**
 * Single-card text IO functions. Later used by cardlist for
 * reading/writing whole hands.
 */
function commonTextOf(n: T, u: boolean): string {
    if (! isCard(n)) return "??";
    if (n === Joker) return "Jk";
    else if (n === RedJoker) return "Jr";
    else if (n === WhiteJoker) return "Jw";

    const r: Rank.T = rankOf(n);
    const s: Suit.T = suitOf(n);
    if (u) return `${Rank.charOf(r)}${Suit.symbolOf(s)}`;
    return `${Rank.charOf(r)}${Suit.charOf(s)}`;
}

export function textOf(n: T): string {
    return commonTextOf(n, false);
}

export function unicodeOf(n: T): string {
    return commonTextOf(n, true);
}

export function unicodeSingleOf(n: number): string {
    if (n < 0 || n > KnightOfSpades) n = 0;
    return [
        "⁇","🃟","🂿","🃏","🃑","🃁","🂱","🂡","🃒","🃂","🂲","🂢","🃓","🃃","🂳","🂣",
        "🃔","🃄","🂴","🂤","🃕","🃅","🂵","🂥","🃖","🃆","🂶","🂦","🃗","🃇","🂷","🂧",
        "🃘","🃈","🂸","🂨","🃙","🃉","🂹","🂩","🃚","🃊","🂺","🂪","🃛","🃋","🂻","🂫",
        "🃝","🃍","🂽","🂭","🃞","🃎","🂾","🂮","🃑","🃁","🂱","🂡","🃜","🃌","🂼","🃜",
    ][n];
}

export function fullNameOf(n: T): string {
    if (! isCard(n)) return "unknown";
    else if (n === Joker) return "joker";
    else if (n === RedJoker) return "red joker";
    else if (n === WhiteJoker) return "white joker";

    const r: Rank.T = rankOf(n);
    const s: Suit.T = suitOf(n);
    return `${Rank.nameOf(r)} of ${Suit.pluralOf(s)}`;
}

export function fromString(s: string): T {
    const [ c, _n ] = parseOne(s, 0);
    return c;
}

/**
 * Simple state machine to parse one card from a piece of text, and give us a
 * pointer past what was read for iterative use.
 */
export function parseOne(text: string, start: number): [T, number] {
    let rank: Rank.T = Rank.None;
    let suit: Suit.T = Suit.None;
    let state: number = 0;
    let done: boolean = false;

    let i = start;
    while (i < text.length) {
        const c = text.charAt(i);

        switch (state) {
        case 0:
            if (" " === c || '[' === c) { break; }
            else if ('J' == c) { state = 2; }
            else {
                rank = Rank.fromChar(c);
                if (rank === Rank.None) return [ None, i ];
                state = 1;
            }
            break;
        case 1:
            suit = Suit.fromChar(c);
            done = true;
            break;
        case 2:
            if ('k' === c) return [ Joker, i + 1 ];
            else if ('r' === c) return [ RedJoker, i + 1 ];
            else if ('w' === c) return [ WhiteJoker, i + 1 ];
            else {
                rank = Rank.Jack;
                suit = Suit.fromChar(c);
                done = true;
            }
            break;
        default:
            throw new Error(`internal`);
        }
        if (done) break;
        i += 1;
    }
    if (suit === Suit.None) return [ None, i ];
    return [ fromRankSuit(rank, suit), i + 1 ];
}

